File: src/main/java/com/kasisoft/libs/common/constants/Digest.java
package com.kasisoft.libs.common.constants;

import static com.kasisoft.libs.common.internal.Messages.*;

import com.kasisoft.libs.common.annotation.*;

import com.kasisoft.libs.common.pools.*;

import com.kasisoft.libs.common.*;

import jakarta.validation.constraints.*;

import java.security.*;

/**
 * Collection of supported MessageDigest implementations.
 *
 * @author daniel.kasmeroglu@kasisoft.com
 */
@Specification(value = "https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/security/MessageDigest.html", date = "14-NOV-2023")
public enum Digest {

    MD2("MD2"),
    MD5("MD5"),
    SHA1("SHA-1"),
    SHA256("SHA-256"),
    SHA384("SHA-384"),
    SHA512("SHA-512");

    private String                  algorithm;
    private Bucket<MessageDigest>   bucket;

    /**
     * Initializes this digest for the supplied algorithm.
     *
     * @param algorithm
     *            The name of the hash algorithm.
     * @throws KclException
     *             The supplied alorithm isn't known.
     */
    Digest(@NotBlank String algorithm) {
        this.algorithm = algorithm;
        // make sure to verify the existence of the algorithm
         getInstance();
    }

    @NotBlank
    public String getAlgorithm() {
        return algorithm;
    }

    @Override
    public String toString() {
        return algorithm;
    }

    private synchronized Bucket<MessageDigest> bucket() {
        if (bucket == null) {
            bucket = new Bucket<MessageDigest>(this::getInstance, MessageDigest::reset);
        }
        return bucket;
    }

    @NotNull
    private MessageDigest getInstance() {
        try {
            return MessageDigest.getInstance(algorithm);
        } catch (NoSuchAlgorithmException ex) {
            // shouldn't happen but better be safe
            throw new KclException(ex, error_unknown_digest_algorithm.formatted(algorithm));
        }
    }

    /**
     * Processes the supplied data blocks in order to calculate a hash.
     *
     * @param data
     *            The data used to be digested.
     * @return The hash value.
     */
    @NotNull
    public String digestToString(@NotNull byte[] ... data) {
        return digestToString(1, data);
    }

    /**
     * Processes the supplied data blocks in order to calculate a hash.
     *
     * @param count
     *            The number of times used to run the digestion.
     * @param data
     *            The data used to be digested.
     * @return The hash value.
     */
    @NotNull
    public String digestToString(int count, @NotNull byte[] ... data) {
        return Buckets.stringBuilder().forInstance($ -> {
            var checksum = digest(count, data);
            for (var b : checksum) {
                var asbyte = Integer.toString((b & 0xff), 16);
                if (asbyte.length() == 1) {
                    $.append('0');
                }
                $.append(asbyte);
            }
            return $.toString();
        });
    }

    /**
     * Processes the supplied data blocks in order to calculate a hash.
     *
     * @param data
     *            The data used to be digested.
     * @return The hash value.
     */
    @NotNull
    public byte[] digest(@NotNull byte[] ... data) {
        return digest(1, data);
    }

    /**
     * Processes the supplied data blocks in order to calculate a hash.
     *
     * @param count
     *            The number of times used to run the digestion.
     * @param data
     *            The data used to be digested.
     * @return The hash value.
     */
    @NotNull
    public byte[] digest(@Min(1) int count, @NotNull byte[] ... data) {
        return bucket().forInstance($digest -> {
            for (var i = 0; i < count; i++) {
                for (var j = 0; j < data.length; j++) {
                    $digest.update(data[j]);
                }
            }
            return $digest.digest();
        });
    }

} /* ENDCLASS */

File: src/main/kotlin/com/kasisoft/libs/common/constants/Digest.kt
package com.kasisoft.libs.common.constants

import com.kasisoft.libs.common.annotation.*

import com.kasisoft.libs.common.pools.*

import com.kasisoft.libs.common.*

import javax.validation.constraints.*

import java.security.*

import java.util.function.Consumer
import java.util.function.Supplier
import java.util.*

/**
 * Collection of supported MessageDigest implementations.
 *
 * @author daniel.kasmeroglu@kasisoft.com
 */
@Specification(value = "https://docs.oracle.com/javase/10/docs/api/java/security/MessageDigest.html", date = "04-JUN-2020")
class Digest(val algorithm : String) {

    private val bucket : Bucket<MessageDigest>

    init {

        try {
            MessageDigest.getInstance(algorithm);
        } catch (ex : NoSuchAlgorithmException) {
            throw KclException(ex, "Unknown digest algorithm '%s'!", algorithm);
        }

        bucket = Bucket<MessageDigest>( Supplier<MessageDigest> { createDigest(algorithm) }, Consumer<MessageDigest> { resetDigest(it) } )

        synchronized (DigestCache) {
            DigestCache.DIGESTS.put(algorithm, this)
        }

    }

    companion object {

        @JvmField val MD2       = Digest("MD2")
        @JvmField val MD5       = Digest("MD5")
        @JvmField val SHA1      = Digest("SHA-1")
        @JvmField val SHA256    = Digest("SHA-256")
        @JvmField val SHA384    = Digest("SHA-384")
        @JvmField val SHA512    = Digest("SHA-512")

        @JvmStatic
        fun values() : List<Digest> {
            synchronized (DigestCache) {
                return DigestCache.DIGESTS.values.toList()
            }
        }

        /**
         * This helper function identifies the MessageDigest algorithm value which corresponds to the supplied name. Be
         * aware that this enumeration only supports the <b>required</b> digests.
         *
         * @param name   The name of the digest which has to be identified. Case sensitivity doesn't matter here.
         *
         * @return   The digest value or empty if it cannot be identified.
         */
        @JvmStatic
        fun findByName(name : String) : Digest? = DigestCache.DIGESTS.get(name)

        fun createDigest(algorithm : String) : MessageDigest? {
            try {
                return MessageDigest.getInstance(algorithm);
            } catch (ex : NoSuchAlgorithmException) {
                // won't happen as this will be checked by the instantiation
                return null;
            }
        }

        fun resetDigest(digest : MessageDigest) : MessageDigest {
            digest.reset();
            return digest;
        }

    } /* ENDOBJECT */

    /**
     * Processes the supplied data blocks in order to calculate a hash.
     *
     * @param data   The data used to be digested.
     *
     * @return   The hash value.
     */
    fun digestToString(vararg data : ByteArray) = digestToString(1, *data)

    /**
     * Processes the supplied data blocks in order to calculate a hash.
     *
     * @param count   The number of times used to run the digestion.
     * @param data    The data used to be digested.
     *
     * @return   The hash value.
     */
    fun digestToString(count : Int, vararg data : ByteArray) : String {
        return Buckets.stringFBuilder().forInstance {
            var checksum = digest(count, *data)
            for (b in checksum) {
                it.appendF("%02x", b)
            }
            it.toString()
        }
    }

    /**
     * Processes the supplied data blocks in order to calculate a hash.
     *
     * @param data   The data used to be digested.
     *
     * @return   The hash value.
     */
    fun digest(vararg data : ByteArray) = digest(1, *data)

    /**
     * Processes the supplied data blocks in order to calculate a hash.
     *
     * @param count   The number of times used to run the digestion.
     * @param data    The data used to be digested.
     *
     * @return   The hash value.
     */
    fun digest(@Min(1) count : Int, vararg data : ByteArray) : ByteArray {
        val digest = bucket.allocate();
        for (i in 0..count - 1) {
            for (j in 0..data.size - 1) {
                digest.update(data[j]);
            }
        }
        val result = digest.digest();
        bucket.free(digest);
        return result;
    }

    override fun toString() = algorithm;

    override fun equals(other : Any?) = (other is Digest) && (algorithm == other.algorithm)

    override fun hashCode() = algorithm.hashCode()

} /* ENDCLASS */

private object DigestCache {

     val DIGESTS = mutableMapOf<String, Digest>()

} /* ENDOBJECT */
